File: w3jl/src/main/java/com/midsummer/w3jl/test/W3JL.java
package com.midsummer.w3jl.test;

import android.util.Log;

import org.web3j.crypto.Credentials;
import org.web3j.crypto.ECKeyPair;
import org.web3j.crypto.Hash;
import org.web3j.crypto.MnemonicUtils;
import org.web3j.crypto.WalletUtils;
import org.web3j.protocol.Web3j;
import org.web3j.protocol.Web3jFactory;
import org.web3j.protocol.core.DefaultBlockParameterName;
import org.web3j.protocol.core.methods.response.EthGetBalance;
import org.web3j.protocol.http.HttpService;
import org.web3j.utils.Convert;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;
import java.util.Random;
import java.util.concurrent.ExecutionException;

/**
 * Created by NienLe on 13,August,2018
 * Midsummer.
 * Ping me at nienbkict@gmail.com
 * Happy coding ^_^
 */
public class W3JL {



    private Web3j web3j;


    public void initWeb3j(String provider){
        web3j = Web3jFactory.build(new HttpService(provider));
    }

    public Web3j getWeb3j() {
        return web3j;
    }

    public String getBalance(String address) throws ExecutionException, InterruptedException {
        EthGetBalance ethGetBalance = web3j
                .ethGetBalance("0xB89d25B8378d8E8b2CB11E6e1bF80fBf33386f45", DefaultBlockParameterName.LATEST)
                .sendAsync()
                .get();
        BigInteger wei = ethGetBalance.getBalance();
        return weiToEth(wei,5);

    }

    public String getMnemonic(){
        byte[] b = new byte[16];
        new Random().nextBytes(b);
        return MnemonicUtils.generateMnemonic(b);
    }


    private static String weiInEth  = "1000000000000000000";

    public static BigDecimal weiToEth(BigInteger wei) {
        return Convert.fromWei(new BigDecimal(wei), Convert.Unit.ETHER);
    }

    public static String weiToEth(BigInteger wei, int sigFig) {
        BigDecimal eth = weiToEth(wei);
        int scale = sigFig - eth.precision() + eth.scale();
        BigDecimal eth_scaled = eth.setScale(scale, RoundingMode.HALF_UP);
        return eth_scaled.toString();
    }

    public String MnemonicToWallet(String mnemonics, String password){
        byte[] seeds = MnemonicUtils.generateSeed(mnemonics, password);
        ECKeyPair ecKeyPair = ECKeyPair.create(Hash.sha256(seeds));
        String prvateKey =ecKeyPair.getPrivateKey().toString(16);
        Credentials credentials = Credentials.create(prvateKey);
        Log.d("MAINNNN","private key: " + prvateKey);
        String address = credentials.getAddress();
        return address;
    }


    public String privateKeyToMnemonic(String privateKey){

        return "";
    }
}

File: w3jl/src/main/java/com/midsummer/w3jl/service/W3JL.kt
package com.midsummer.w3jl.service

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.google.gson.Gson
import com.midsummer.w3jl.entity.W3JLBip39Wallet
import com.midsummer.w3jl.entity.W3JLCredential
import com.midsummer.w3jl.entity.W3JLWallet
import com.midsummer.w3jl.util.FileUtil
import io.github.novacrypto.bip39.MnemonicGenerator
import io.github.novacrypto.bip39.Words
import io.github.novacrypto.bip39.wordlists.English
import org.web3j.crypto.*
import org.web3j.protocol.Web3j
import org.web3j.protocol.core.DefaultBlockParameterName
import org.web3j.utils.Numeric
import java.io.File
import java.io.IOException
import java.math.BigInteger
import java.security.SecureRandom
import java.util.*

/**
 * Created by NienLe on 14,August,2018
 * Midsummer.
 * Ping me at nienbkict@gmail.com
 * Happy coding ^_^
 */
class W3JL(var web3j: Web3j, var filePath: File) : W3JLRepository{
    private val RADIX = 16
    private val GAS_PRICE = BigInteger.valueOf(20_000_000_000L)
    private val GAS_LIMIT = BigInteger.valueOf(4300000)
    private val objectMapper = jacksonObjectMapper()


    override fun createBip39WalletWithPassword(password: String): W3JLBip39Wallet {
        return createBip39WalletWithPasswordAndMnemonic(password, createMnemonics())
    }

    override fun createBip39WalletWithMnemonic(mnemonics: String): W3JLBip39Wallet {
        return createBip39WalletWithPasswordAndMnemonic("", mnemonics)
    }

    override fun createBip39WalletWithPasswordAndMnemonic(password: String, mnemonics: String): W3JLBip39Wallet {
        val seed = MnemonicUtils.generateSeed(mnemonics, password)
        val keyPair = ECKeyPair.create(Hash.sha256(seed))
        val walletFile = WalletUtils.generateWalletFile(password, keyPair, filePath, false)
        return W3JLBip39Wallet(walletFile, mnemonics, keyPair.privateKey.toString(RADIX), keyPair.publicKey.toString(RADIX),
                getAddressFromPrivateKey(keyPair.privateKey.toString(RADIX)))
    }

    override fun createMnemonics(entropy: ByteArray): String {
        val sb = StringBuilder()
        SecureRandom().nextBytes(entropy)
        MnemonicGenerator(English.INSTANCE)
                .createMnemonic(
                        entropy
                ) {
                    sb.append(it)
                }
        return sb.toString()
    }

    override fun createMnemonics(): String {
        return createMnemonics(ByteArray(Words.TWELVE.byteLength()))
    }

    override fun mnemonicsToPrivateKey(mnemonics: String, password: String?): String {
        return mnemonicsToKeyPair(mnemonics, password).privateKey.toString(RADIX)
    }

    override fun mnemonicsToPublicKey(mnemonics: String, password: String): String {
        return mnemonicsToKeyPair(mnemonics, password).publicKey.toString(RADIX)
    }

    override fun mnemonicsToKeyPair(mnemonics: String, password: String?): ECKeyPair {
        val seeds = MnemonicUtils.generateSeed(mnemonics, password)
        return ECKeyPair.create(Hash.sha256(seeds))
    }

    override fun walletToKeyPair(wallet: Bip39Wallet, password: String): ECKeyPair {
        return mnemonicsToKeyPair(wallet.mnemonic, password)
    }

    @Throws(IOException::class)
    override fun loadCredential(password: String, file: File): W3JLCredential? {
        val c = WalletUtils.loadCredentials(password, file)
        return W3JLCredential(c.address, c.ecKeyPair.publicKey.toString(RADIX), c.ecKeyPair.privateKey.toString(RADIX))
    }

    override fun loadCredentialFromPrivateKey(privateKey: String): Credentials? {
        return if (!WalletUtils.isValidPrivateKey(privateKey)) {
            null
        } else Credentials.create(privateKey)
    }

    override fun getAddressFromPrivateKey(privateKey: String): String {
        val credentials = loadCredentialFromPrivateKey(privateKey) ?: return ""
        return credentials.address
    }

    override fun transfer(password: String, walletFile: File, from: String, to: String, amount: BigInteger): String? {
        val credentials = WalletUtils.loadCredentials(password, walletFile) ?: return null

        val ethGetTransactionCount = web3j.ethGetTransactionCount(
                from, DefaultBlockParameterName.LATEST).sendAsync().get()

        val nonce = ethGetTransactionCount.transactionCount
        println(nonce)

        val rawTransaction = RawTransaction.createEtherTransaction(
                nonce, GAS_PRICE, GAS_LIMIT, to, amount)
        val signedMessage = TransactionEncoder.signMessage(rawTransaction, credentials)
        val hexValue = Numeric.toHexString(signedMessage)

        val ethSendTransaction = web3j.ethSendRawTransaction(hexValue).sendAsync().get()
        return ethSendTransaction.transactionHash
    }

    override fun transfer(privateKey: String, from: String, to: String, amount: BigInteger): String? {
        val credentials = loadCredentialFromPrivateKey(privateKey) ?: return null

        val ethGetTransactionCount = web3j.ethGetTransactionCount(
                from, DefaultBlockParameterName.LATEST).sendAsync().get()

        val nonce = ethGetTransactionCount.transactionCount
        println(nonce)

        val rawTransaction = RawTransaction.createEtherTransaction(
                nonce, GAS_PRICE, GAS_LIMIT, to, amount)
        val signedMessage = TransactionEncoder.signMessage(rawTransaction, credentials)
        val hexValue = Numeric.toHexString(signedMessage)

        val ethSendTransaction = web3j.ethSendRawTransaction(hexValue).sendAsync().get()
        return ethSendTransaction.transactionHash
    }


    override fun createWalletFromMnemonic(mnemonics: String, password: String?): W3JLWallet {
        val seed = MnemonicUtils.generateSeed(mnemonics, password)
        val keyPair = ECKeyPair.create(Hash.sha256(seed))
        val tmpWallet = Wallet.createLight(password, keyPair)
        val wallet = W3JLWallet()
        wallet.mnemonic = mnemonics
        wallet.source = W3JLWallet.Source.MNEMONIC
        wallet.address = "0x${tmpWallet.address}"
        wallet.privateKey = keyPair.privateKey.toString(RADIX)
        wallet.jsonSource = Gson().toJson(tmpWallet)
        wallet.createAt = Calendar.getInstance().timeInMillis
        return wallet
    }

    override fun createWalletFromPrivateKey(privateKey: String, password: String?): W3JLWallet {
        val wallet = W3JLWallet()
        if (!WalletUtils.isValidPrivateKey(privateKey)){
            return wallet
        }
        wallet.source = W3JLWallet.Source.PRIVATE_KEY
        val c = Credentials.create(privateKey)
        val tmpWallet = Wallet.createLight(password, c.ecKeyPair)
        wallet.address = c.address
        wallet.privateKey = privateKey
        wallet.jsonSource = objectMapper.writeValueAsString(tmpWallet)
        wallet.createAt = Calendar.getInstance().timeInMillis
        return wallet
    }

    override fun createWalletFromJsonString(jsonString: String, password: String?): W3JLWallet {
        val walletFile : WalletFile = objectMapper.readValue(jsonString)
        val c = Credentials.create(Wallet.decrypt(password, walletFile))
        val wallet = W3JLWallet()
        wallet.source = W3JLWallet.Source.JSON
        wallet.address = c.address
        wallet.privateKey = c.ecKeyPair.privateKey.toString(RADIX)
        wallet.jsonSource = jsonString
        wallet.createAt = Calendar.getInstance().timeInMillis
        return wallet
    }
}